File: lua/gw_reg.lua
require "stringadd"

local execcmd = function(command)
   local handle = io.popen(command, "r")
   local content = handle:read("*all")
   handle:close()
   return content
end

local ucs2c = function(ucs)
   --print("??",ucs);
   local hex = ucs:match("u(%x+)");
   if (not hex) then return "〓"; end
   return utf8.char(tonumber("0x"..hex));
end

local getjson = function(page)
   local cmd = "curl -w '%{http_code}' http://glyphwiki.org/json?name=" .. page;
   cmd = cmd .. " -s -S"
   print(cmd);
   local ret = execcmd(cmd);
   local m = ret:match('"data":%"([^%"]+)%"');
   local c = ret:match('"related":%"U%+([^%"]+)%"');
   if (c) then
      c = tonumber(c, 16);
      if (true or c < 0x10000) then
         c = utf8.char(c);
      else
--         c = c - 0x10000;
--         c = utf8.char((0xD800 | (c >> 10)), (0xDC00 | (c & 0x3FF)));
      end
   end
   return m, c;
end

--[[
local getjson = function(page)
   local ret = execcmd("grep '^ " .. page .. " ' ../gwdiff/cgi/gw1205/dump_newest_only.txt  -r");
   local data = ret:split("|");
   if (#data < 3) then return nil, nil end

   local m = data[3]:trim();
   local c = data[2]:trim();

   if (c) then
      c = ucs2c(c);
   end
   
   --print ("current = ",m,c);
   return m,c;
   end
]]
local accesswiki = function(page, glyph, c, overwrite)
   if (false) then
      local fp = io.open("reglog/"..page..".dat","r")
      if (fp) then
         fp:close();
         return
      end
   end

   print("#####success", page, c, glyph, overwrite);

   if (arg[1] ~= "exist" and arg[1] ~= "register") then
      return
   end


   -- 入れ替えの場合
   if (overwrite == "swap") then
      local cmd = "curl -w '\\n'"
      cmd = cmd .. (" 'http://glyphwiki.org/wiki/%s?action=swap'"):format(page);
      cmd = cmd .. " -b gwcookie.txt"
      print(cmd);
      
      if (arg[1] == "register") then
         execcmd(cmd);
         --execcmd("sleep 1");
      end
      return;
   end

   local summ = nil;
   -- check if there's registered glyph
   local data,c0 = getjson(page);

   -- 関連字だけ差し替える場合
   if (overwrite == "kr") then
      glyph = data;
      summ = "関連字";

   -- 現在最新が実体である場合
   elseif (data and not data:match("^99:0:0:0:0:200:200:([^:]+)$")) then

      -- 自分を参照しているやつらを拾い出す
      local ret = execcmd("grep '99:0:0:0:0:200:200:" .. page .. "$' ../gwdiff/cgi/gw1205/dump_newest_only.txt  -r");
      local k = ret:split("\n");
      
      -- 上書きの場合(参照がいれば上書きしない:要swap)
      if (overwrite == "overwrite") then
         if (0 < #k) then
            -- todo: swap
            local cmd = "echo " .. page .. " as otheralias >> ./reglog/0undone ";
            execcmd(cmd);
            return
         end
      -- 両保持の場合(参照がいればvarを検索)
      elseif (overwrite == "branch")  then
         local ret = execcmd("grep '99:0:0:0:0:200:200:" .. page .. "$' ../gwdiff/cgi/gw1205 -r");
         if (#k == 0) then
            -- todo: "u%x-var-%d":format(c, i) の作成
            local ret = execcmd("grep '^ " .. ("u%x-var-"):format(utf8.codepoint(c)) .. "' ../gwdiff/cgi/gw1205/dump_newest_only.txt  -r");
            local n = ret:split("\n");

            local cmd = "echo '" .. page .. " as nobranch ";
            cmd = cmd .. ("(u%x-var-%03d)"):format(utf8.codepoint(c), #n + 1);
            cmd = cmd .. "' >> ./reglog/0undone ";
            execcmd(cmd);
            return;
         end
         local page = k[1]:split(" ");
         execcmd("echo 'https://glyphwiki.org/wiki/".. page[1] .."?action=swap' >> ./reglog/0undone");

         return;
      else
         local cmd = "echo " .. page .. " as original >> ./reglog/0undone ";
         execcmd(cmd);
         return
      end
   end

   if (glyph:match("^%[%[.+%]%]$")) then
      summ = glyph;
   end

   local timestump = os.time();

   local cmd = "curl -w '\\n'"
   cmd = cmd .. (" 'http://glyphwiki.org/wiki/%s'"):format(page);
   cmd = cmd .. (" -d 'page=%s' "):format(page);
   cmd = cmd .. (" -d 'edittime=%d' "):format(timestump);
   cmd = cmd .. (" -d 'textbox=%s'"):format(glyph);
   if (c)  then cmd = cmd .. (" -d 'related=%s'"):format(c); end
   if (summ) then cmd = cmd .. (" --data-urlencode 'summary=%s'"):format(summ); end
   cmd = cmd .. (" --data-urlencode 'buttons=以上の記述を完全に理解し同意したうえで投稿する'");
   cmd = cmd .. " -b gwcookie.txt"
   cmd = cmd .. (" >> ./reglog/%s.dat"):format(page);

   print(cmd);

   if (arg[1] == "register") then
      execcmd(cmd);
      execcmd("sleep 1");
   end
end

local loginwiki = function()

   local cmd = "curl -w '\\n'"
      .. " http://glyphwiki.org/wiki/Special:Userlogin"
      .. " -d 'action=page'"
      .. " -d 'buttons=ログイン'"
      .. " -d 'name=mtnest'"
      .. " -d 'page=Special:Userlogin'"
      .. " -d 'password=s57sep'"
      .. " -d 'returnto=Special:Userlogout'"
      .. " -d 'type=login'"
      .. " -b gwcookie.txt"
      .. " -c gwcookie.txt"
   print(cmd);
   local t = execcmd(cmd);
   print(t);
   return t:find("ログイン成功");
end

local glyphname = function(code)
   code = code:lower()
   if (code:match("^dh*%d+%.%d$")) then
      return "dkw-" .. code:sub(2, 6) .. ("d"):rep(tonumber(code:sub(-1)));
   end

   local name = ""
   for w in code:gmatch("u%+%x+") do
      name = name .. "-u" .. ("%x"):format(tonumber(w:sub(3), 16))
   end
   return name:sub(2)
end


local uk2d;
local d2uk;
local ucs2dkw = function(key)
   if (uk2d ~= nil) then
      return uk2d[key];
   end
   uk2d = {};
   d2uk = {};
   local fp = io.open("mji_00502_pickup.txt","r")
   for gline in fp:lines() do
      local line = gline:split("\t");
      local dkw = line[1];
      local koseki = line[2];
      local ucs = line[3]:lower():gsub("^u%+", "u");
      local ivs0 = (line[4] or ""):lower();
      local ivs = (ivs0:match("^%x+_%x+$")) and "u" .. ivs0:gsub("_", "-u") or ucs;
      local page = (function(dkw)
        local page = dkw:trim():gsub("補", "h"):gsub("%'", "d");
        local n = page:match("%d+")
        return page:gsub(n, (page:find("^h") and "%04d" or "%05d"):format(n));
      end)(dkw);

      local alias = (function(koseki)
         local glyph = koseki:match("^ksk%-(%d+)$");
         return glyph and ("koseki-%s"):format(glyph) or nil;
      end)(koseki);
      uk2d[ucs] = uk2d[ucs] and uk2d[ucs] .. "," .. page or page;
      if (ivs0:match("^u%+f[9a]") or ivs0:match("^u%+2f")) then uk2d[ivs0:gsub("u%+","u")] = page; end
      
      if (alias) then
         uk2d[alias] = page;
         d2uk[page] = {koseki=alias, c=ivs};
      else
         d2uk[page] = {koseki="jmj",c=ivs};
      end
   end
   fp:close();
   return uk2d[key];
end


--版1: dkw2ucs.txtのとおりに登録する
local dkw2ucs = function(line)
    if (line == "") then return; end 
    local data = line:split("#");
    if (data[2]) then
       if (data[2]:find("removed") or
           data[2]:find("missing") or
           data[2]:find("moved")) then
          return;
       end
   end

   local tbl = data[1]:trim():split(" ");
   local code = tbl[1];

   if (code < "D19090") then return; end
   if ("D19100" < code) then return; end

   local page = glyphname(code)
   local alias = "[[" .. glyphname(tbl[5]) .. "]]"
   accesswiki(page, alias);
   return;
end

--版1: dkw2ucs.txtのとおりに登録する
local crawl_table = function()
   local fp = io.open("dkw2ucs.txt","r")
   for line in fp:lines() do
      dkw2ucs(line);
   end
   fp:close();
end

--版2: mj.00502.xmlから直接登録する
local pickups = function(buff)
   local dkw = buff:match("<大漢和>([^<]+)</大漢和>");
   if (not dkw) then return end
   local ucs = buff:match("<対応するUCS[^>]->([^<]+)</対応するUCS>");
   local font = buff:match("<実装した[^>]->([^<]+)</実装した");
   local koseki = buff:match("<戸籍統一文字番号([^<]+)<");
   if (font == nil or ucs==font) then  font = "" end
   if (ucs:sub(3) == font:sub(1, ucs:len() - 2)) then
      font = "U+" .. font:gsub("_", "U+")
      ucs = ""
   end
   --print(dkw, koseki, font, ucs);
   local page = (function(dkw)
         local page = dkw:gsub("補", "h"):gsub("%'", "d");
         local n = page:match("%d+")
         return "dkw-" .. page:gsub(n, (page:find("^h") and "%04d" or "%05d"):format(n));
   end)(dkw);

   local alias = (function(koseki)
         local glyph = koseki:match("^%>(%d+)$");
         return glyph and ("[[koseki-%s]]"):format(glyph) or nil;
   end)(koseki);
   if (alias) then
      accesswiki(page, alias);
   end
end

--版2: mj.00502.xmlから直接登録する
local crawl_table = function(e0, e1)
   local fp = io.open("./mji.00502/mji.00502.xml")

   local buff = "";
   local i = 1;
   for line in fp:lines() do
      if (line:find("<MJ文字情報>")) then buff = ""; end
      buff = buff .. line:trim();
      if (line:find("</MJ文字情報>")) then
         if (e0 <= i) then
            pickups(buff);
         end
         i = i + 1;
         if (e1 < i) then break; end
      end
   end
end

--版3: TSVから間接登録する
local pickups = function(buff)
   line = buff:split("\t");
   local dkw = line[1];
   local koseki = line[2];
   local incl = line[4];
   local page = glyphname(dkw);
   local alias = (function(koseki)
      local glyph = koseki:match("^%>(%d+)$");
      return glyph and ("[[koseki-%s]]"):format(glyph) or nil;
   end)(koseki);

   -- ついでにreplaceもしていたが複雑すぎるので分離
   if (false) then
      if (incl ~= "/" and alias) then
         local parts = incl2array(incl);
         if (parts == nil) then return; end
         
         --登録済みチェックはここでやる
         local cmd = "curl -w '%{http_code}' http://glyphwiki.org/glyph/" .. page .. ".50px.png -o /dev/null -s"
      
         if (execcmd(cmd) ~= "404") then
            return;
         end

         return replace_parts(alias:sub(3,-3),parts, page);
      end
   end

   if(true) then
      if (alias == nil) then return end
      accesswiki(page, alias);
   end
end

--版3: TSVから間接登録する
local crawl_table = function(e0, e1)
   local fp = io.open("../kids/mj0502_pickup2.txt","r"); --のちのmj1_ishii.txt
   local i = 0;
   for line in fp:lines() do
      i = i + 1;
      if (e0 <= i) then pickups(line) end
      if (e1 <= i) then break end
   end
   fp:close();
end

local ucs2c = function(ucs)
   --print("??",ucs);
   local hex = ucs:match("u(%x+)");
   if (not hex) then return ucs; end
   return utf8.char(tonumber("0x"..hex));
end

--版4: reg.datのとおりに登録する
local pickups = function(buff)
   line = buff:split("\t");
   if (#line < 2) then  return; end
   local dkws = line[1]:split("[");
   local dkw = dkws[1];
   local c = (dkws[2] or "〓"):split("]")[1] or "〓";
   local glyph = line[2];
   local cat = line[3];
   local ow = line[4] or "overwrite"
   
   
   if (dkw:match("^dkw%-")==nil and dkw:match("^u%x") == nil
         and dkw:match("^cdp%-") == nil 
          and dkw:match("^nyukan%-")==nil and dkw:match("^toki%-") == nil
          and dkw:match("^gt%-")==nil and dkw:match("^jmj%-") == nil
   ) then
      print("##format");
      return
   end
   if (not glyph) then
      print("##no glyph");
      return
   end
   if (c:match("^u%x+")) then
      local ucs = "0x" .. c:match("^u(%x+)");
      c = utf8.char(tonumber(ucs));
   end
   glyph = table.join(glyph:split("$"), "$");
   if (glyph:sub(1,2) == "[[") then
      c = "〓"
   end

   if (cat:match("^#[hr%-]") or glyph == "_" or glyph == "[[]]") then
      local cmd = "echo " .. dkw .. " as undef >> ./reglog/0undone ";
      execcmd(cmd);
      return
   end
   accesswiki(dkw, glyph, c, ow);

end

--版4: reg.datのとおりに登録する
local crawl_table = function(fname, e0, e1)
   local fp = io.open(fname,"r")
   --local fp = io.open("reglog/swaps","r")
   local i = 0;
   for line in fp:lines() do
      i = i + 1;
      if (e0 <= i) then pickups(line) end
      if (e1 <= i) then break end
   end
   fp:close();
end

local help = function()
   print (" # " .. arg[0] .. " [mode] [file] [start] [end]");
   print ("");
   print (" [mode] ");
   print ("  - dump: Only dump");
   print ("  - login: Login glyphwiki test");
   print ("  - exist: Access glyphwiki and dump yet-registered glyphs");
   print ("  - register: registration (with login and exist) ");
   print ();
   print (" [start][end]");
   print ("  - xml mji entry");
   print ();
end

if (#arg ~= 4) then
   help();
   return;
end

execcmd("mkdir -p ./reglog");

if ((arg[1] == "login" or (arg[1] == "register")) and (loginwiki() == nil)) then
   return;
end


local e0 = tonumber(arg[3]);
local e1 = tonumber(arg[4]);
if (e0 > e1) then
   help();
   return;
end
crawl_table(arg[2], e0, e1);


if (false) then
   local glyph = "99:0:0:0:0:150:200:u72ad-01:0:0:0$99:0:0:60:0:199:95:u8278:0:0:0$99:0:0:65:80:192:190:u7530:0:0:0";
   accesswiki("dkw-11232", "[[u2283a]]");
end

File: lua/gw_reg.js
let arg = process.argv.slice(1);
console.log(arg);

const DUMPNEWEST = "dump_newest_only.txt";
const LOGPATH = "./reglog/";
const GWSITE = "https://glyphwiki.org/wiki/";
const UNDONE_LOG  = "./reglog/0undone";
const COOKIE = LOGPATH + "gwcookie.txt";

let ucs2c = (ucs) => (ucs[0].toLowerCase() == "u")
    ? String.fromCodePoint(parseInt(ucs.slice(1),16) || 0x3013)
    : "〓";

const getfile = (fname) => require('fs').readFileSync(fname, 'utf8');
const { execSync } = require('child_process')
const execcmd = (cmd) => execSync(cmd + "; true").toString();
const undone = (message) => arg[1] == "register" ? execcmd(`echo '${message}' >> ${UNDONE_LOG}`) : console.log("undone: " + message);

let getjson = function(page) {
    let cmd = "curl https://glyphwiki.org/json?name=" + page;
    cmd = cmd + " -s -S"
    let ret = execcmd(cmd);
    ret = JSON.parse(ret);
    let m = ret.data;
    let c = ret.related;
    if (c) c = ucs2c(c);
    if (c == "〓") return [m];
    return [m, c];
}

/*
let getjson = function(page) {
    let ret = execcmd("grep '^ ${page} ' ${DUMPNEWEST}");
    let data = ret.split("|");
    if (data.length < 3) return;

   let m = data[2].trim();
   let c = data[1].trim();

   if (c) c = ucs2c(c);
   //console.log("current = ",m,c);
   return m,c;
}
*/
let accesswiki = function(page, glyph, c, overwrite, summ) {
    console.log("#####try:", page, c, overwrite, glyph.slice(0,30) + (30 < glyph.length ? "...": "   "));

   if (arg[1] != "dryrun" && arg[1] != "register") { 
       return;
   }
   // 入れ替えの場合
   if (overwrite == "swap") { 
       let cmd = `curl -w '\\n '${GWSITE}/${page}?action=swap' -b ${COOKIE} `;
       console.log(cmd);
       if (arg[1] == "register") execcmd(cmd);
       return;
   }

    // 現行を確認
    let glyph0, c0;
    [glyph0, c0] = getjson(page);

    if (glyph0 && overwrite == "new") {
        return undone(page + "already registered");
    }

    // 参照
    let toalias = glyph.match(/^\[\[(.+)\]\]$/);
    let toalias0 = glyph0 && glyph0.match(/^99:0:0:0:0:200:200:([^:]+)$/);

    // 関連字だけ差し替える場合
    if (overwrite == "kr") { 
        glyph = glyph0;
        summ = "関連字";

    } else if (glyph0 && !toalias0 && overwrite != "force") {        // 現在最新が実体である場合

        // 自分を参照しているやつらを拾い出す
        let k = execcmd(`grep '99:0:0:0:0:200:200:${page}$' ${DUMPNEWEST}`).trim().split("\n").filter(v => v);

        if (overwrite == "isolated") {
            // 被参照あれば上書きしない, todo:swap
            if (0 < k.length) return undone(page + " as otheralias");
        } else if (overwrite == "alias")  {
            // 被参照があればそれとswap提案
            if (0 < k.length) return undone(`${GWSITE}${k[0].trim().split(" ")[0]}?action=swap`);

            // 被参照がなければ現行を"u%x-var-%d"に改名・swap提案
            let n = execcmd("grep '^ u${c.codePointAt(0).toString(16)}-var-' ${DUMPNEWEST}").split("\n").length;

            let message = `${page} as nobranch `
                + `(u${ c.codePointAt(0).toString(16) }-var-${ ("00" + (n + 1)).slice(-3) })`;
            return undone(message);

        } else {
            return undone(page + " as original");
        }
   }

    // 参照なら概要に記載
    if (!summ && toalias) {
        summ = glyph;
        // ToDo: その日のうちに変更した文字は対象外
        let data = execcmd(`grep '^ ${glyph.slice(2,-2)} ' ${DUMPNEWEST}`).split("|");
        let match = (data[2]||"").trim().match(/^99:0:0:0:0:200:200:([^:]+)$/);
        if (match) summ = "[[" + match[1] + "]]";

    }
    // 関連字
    c = c || c0;
    if (!toalias && !c) {
        // 実体にもかかわらずrelatedの指定がない場合は自動算出する
        let _c = toalias0 && execcmd(`grep '^ ${toalias0[1]} ' ${DUMPNEWEST}`).split("|")[1];
        if (!_c || _c.trim() == "u3013") return undone(`${page} no kanrenji`);
        c = ucs2c(_c.trim());
    }

    let timestamp = parseInt(Date.now()/1000);

    let cmd = `curl -w '\\n' '${GWSITE}/${page}'`
        + ` -d 'page=${page}' `
        + ` -d 'related=${c || "〓"}'`
        + ` -d 'edittime=${timestamp}'`
        + ` -d 'textbox=${glyph}'`;
    if (summ)  cmd += ` -d 'summary=${summ}'`;
    console.log(cmd);

    if (arg[1] == "register") {
        cmd += ` -d 'buttons=以上の記述を完全に理解し同意したうえで投稿する'`
            + ` -b ${COOKIE}`
            + ` >> ${LOGPATH}${page}.dat`;

        execcmd(cmd);
        execcmd("sleep 1");
    }
}

let loginwiki = function() {
    let pwd = getfile(LOGPATH + "pwd").trim();
    let cmd = "curl -w '\\n'"
        + ` ${GWSITE}/Special:Userlogin`
        + " -d 'action=page'"
        + " -d 'buttons=ログイン'"
        + " -d 'name=mtnest'"
        + " -d 'page=Special:Userlogin'"
        + ` -d 'password=${pwd}'`
        + " -d 'returnto=Special:Userlogout'"
        + " -d 'type=login'"
        + ` -b ${COOKIE}`
        + ` -c ${COOKIE}`;
    console.log(cmd);
    let t = execcmd(cmd);
    console.log(t);
    return t.match("ログイン成功");
}



//版4: reg.datのとおりに登録する
let pickups = function(buff) {
    let dkws, glyph, cat, ow, dkw, c;
    [dkws, glyph, cat, ow] = buff.split("\t");
    if (!dkws) return;
    [dkw, c] = dkws.split("[");
    c = (c || "").split("]")[0];
    ow = ow || "isolated";

    if (!dkw.match(/^u[0-9a-f]/) &&
        "dkw,cdp,nyukan,gt,jmj,toki,koseki".split(",").every(prefix => dkw.indexOf(prefix + "-") != 0)
       ) {
        console.log("##format:", dkw);
        return;
    }
    if (!glyph) {
        console.log("##no glyph:", dkw);
        return;
    }
    if (c.match(/^u([0-9a-f]+)/)) c = ucs2c(c);
    glyph = glyph.split("$").filter(v => v).join("$");
    if (!c) c = null;
    //if (!c && glyph.slice(0,2) == "[[") c = null;

    // #ho, #refは登録しない
    if (cat.match(/^#[hrs\-]/) || glyph == "_" || glyph == "[[]]") { 
        let message = `${dkw} as undef ` + cat;
        return undone(message);
    }
    accesswiki(dkw, glyph, c, ow, cat.match("Revert") && cat.split(":").pop().trim());
}

let help = function() {
    console.log(" # " + arg[0] + " [mode] [file] [start] [end]");
    console.log("");
    console.log(" [mode] ");
    console.log("  - dump: Only dump");
    console.log("  - login: Login glyphwiki test");
    console.log("  - dryrun: Access glyphwiki and dump yet-registered glyphs");
    console.log("  - register: registration (with login and dryrun) ");
    console.log();
    console.log(" [file]");
    console.log("  the path of tsv file as the following format");
    console.log("    1st column: dkw-00001[u4e00] (for example of glyphname[related])");
    console.log("    2nd column: [[u4e00-j]]      (for example of glyph data)");
    console.log("    3nd column: #a               (for example of category)");
    console.log("    4th column: (one of the following, 'isolated' by default)")
    console.log("      - force    ; overwrite anyway");
    console.log("      - isolated ; overwrite only if alias or not referred to by other glyphs");  // overwrite
    console.log("      - alias    ; overwrite only if alias (otherwise suggest rename)"); // branch
    console.log("      - new      ; register only if new"); // onlynew
    console.log("      - kr       ; register only related");
    console.log("      - swap     ; swap as substance");
    console.log();
    console.log(" [start][end]");
    console.log("  tsv lines");
    console.log();
}

if (arg.length != 5) { 
   help();
   return;
}

execcmd(`mkdir -p ${LOGPATH}`);

if ((arg[1] == "login" || (arg[1] == "register")) && !loginwiki()) { 
    return;
}

let e0 = parseInt(arg[3]);
let e1 = parseInt(arg[4]);
let fname = arg[2];
if (e0 > e1) return help();
getfile(fname).split("\n").slice(e0,e1).map(line => pickups(line));

if (false) { 
    let glyph = "99:0:0:0:0:150:200:u72ad-01:0:0:0$99:0:0:60:0:199:95:u8278:0:0:0$99:0:0:65:80:192:190:u7530:0:0:0";
    accesswiki("dkw-49181", "[[u23727-jv]]", ucs2c("u23727"), "alias");
}
